internal/datanode/data_node.go

```diff
@@ -547,48 +547,51 @@ func (node *DataNode) FlushSegments(ctx context.Context, req *datapb.FlushSegmen
 		fmt.Sprint(Params.DataNodeCfg.GetNodeID()),
 		MetricRequestsTotal).Inc()
 
-	status := &commonpb.Status{
+	errStatus := &commonpb.Status{
 		ErrorCode: commonpb.ErrorCode_UnexpectedError,
 	}
 
 	if node.State.Load().(internalpb.StateCode) != internalpb.StateCode_Healthy {
-		status.Reason = "DataNode not in HEALTHY state"
-		return status, nil
+		errStatus.Reason = "dataNode not in HEALTHY state"
+		return errStatus, nil
 	}
 
-	log.Info("Receive FlushSegments req",
-		zap.Int64("collectionID", req.GetCollectionID()),
+	log.Info("receiving FlushSegments request",
+		zap.Int64("collection ID", req.GetCollectionID()),
 		zap.Int64s("segments", req.GetSegmentIDs()),
 		zap.Int64s("stale segments", req.GetMarkSegmentIDs()),
 	)
 
-	processSegments := func(segmentIDs []UniqueID, flushed bool) bool {
+	// TODO: Here and in other places, replace `flushed` param with a more meaningful name.
+	processSegments := func(segmentIDs []UniqueID, flushed bool) ([]UniqueID, bool) {
 		noErr := true
-		for _, id := range segmentIDs {
-			if node.segmentCache.checkIfCached(id) {
-				// Segment in flushing, ignore
-				log.Info("segment flushing, ignore the flush request until flush is done.",
-					zap.Int64("collectionID", req.GetCollectionID()), zap.Int64("segmentID", id))
-				status.Reason = "segment is flushing, nothing is done"
-				noErr = false
+		var flushedSeg []UniqueID
+		for _, segID := range segmentIDs {
+			// if the segment in already being flushed, skip it.
+			if node.segmentCache.checkIfCached(segID) {
+				log.Info("segment is already being flushed, ignoring flush request",
+					zap.Int64("collection ID", req.GetCollectionID()),
+					zap.Int64("segment ID", segID))
 				continue
 			}
-
-			node.segmentCache.Cache(id)
-
-			flushCh, err := node.flowgraphManager.getFlushCh(id)
+			// Get the flush channel for the given segment ID.
+			// If no flush channel is found, report an error.
+			flushCh, err := node.flowgraphManager.getFlushCh(segID)
 			if err != nil {
-				status.Reason = "no flush channel found for v-channel"
-				log.Error("no flush channel found for v-channel", zap.Error(err))
+				errStatus.Reason = "no flush channel found for the segment, unable to flush"
+				log.Error(errStatus.Reason, zap.Int64("segment ID", segID), zap.Error(err))
 				noErr = false
 				continue
 			}
 
+			// Cache the segment and send it to its flush channel.
+			flushedSeg = append(flushedSeg, segID)
+			node.segmentCache.Cache(segID)
```

Comment:
should we use segmentCache.Cache() and node.segmentCache.checkIfCached() in one call to make it atomically

Reply:
Sure i have made it atomic.